repair/repair.cc

```diff
@@ -1959,23 +1985,66 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             }
             auto& strat = erm->get_replication_strategy();
             dht::token_range_vector ranges = strat.get_ranges(myid, *tmptr).get0();
-            auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            sstring ks_source_dc;
+            // Allow repairing in the source_dc only if there are enough replicas remaining
+            if (!source_dc_live_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    const auto& nt_strat = dynamic_cast<const locator::network_topology_strategy&>(strat);
+                    ssize_t rf = nt_strat.get_replication_factor(source_dc);
+                    size_t nr_racks = source_dc_racks.size();
+                    size_t max_replicas_per_rack = (rf + nr_racks - 1) / nr_racks;
+                    size_t possibly_lost_replicas = 0;
+                    for (const auto& [rack, nr] : lost_nodes_per_rack) {
+                        possibly_lost_replicas += std::min(nr, max_replicas_per_rack);
+                    }
+                    // Can we ensure a quorum in source_dc?
+                    if (possibly_lost_replicas < rf / 2) {
```

Comment:
This assumes user always use quorum read and writes. What if they do not? For example, RF =1 , WRITE = 1 , READ = 3. With 1 node down, sync with local dc nodes could break consistency. E.g., the write (say delete a partition) went to the later down node only. Before the replace, consistency is ok (read will be rejected). After the replace, consistency breaks, data could be  resurrected even if with READ = 3.

Reply:
Btw, the check should be:

if (possibly_lost_replicas <= rf / 2)

.

Reply:
we care about the complimentary part, possibly_lost_replicas==rf/2 means remaining_replicas==rf/2 and that's too little for a quorum.

Reply:
Say we have RF = 3, and 3 racks, each has 1 node. 
When node 1 in rack a is down, possibly_lost_replicas = 1, we still have the quorum. 

But the check below think the quorum is not meet. (1 < 3/2 == false)

```
if (possibly_lost_replicas < rf / 2) {
     source_dc_for_keyspace = source_dc->name;
} else {
}

```

Reply:
> We cannot control the consistency level used by the application, but I think that guaranteeing quorum semantics in the source dc is good enough in the practical sense. A user should be able to run repair on top of them to fix any cross-dc inconsistencies.

Users can always run cross-dc repair to fix the inconsistency after the node operation, but it is too late. That's why we want to make the node operation itself safe without requiring extra repair.

 
Using the logic "rebuild from quorum node" is just ignoring the fact we allow users to use R + W > RF and assuming users always use R=quorum and W= quorum. Alternatively, if we fallback to use all or more than one DC, it is slower but it is safer.

Reply:
> Say we have RF = 3, and 3 racks, each has 1 node. When node 1 in rack a is down, possibly_lost_replicas = 1, we still have the quorum.
> 
> But the check below think the quorum is not meet. (1 < 3/2 == false)
> 
> ```
> if (possibly_lost_replicas < rf / 2) {
>      source_dc_for_keyspace = source_dc->name;
> } else {
> }
> ```

Good point.  Will fix.

Reply:
> > We cannot control the consistency level used by the application, but I think that guaranteeing quorum semantics in the source dc is good enough in the practical sense. A user should be able to run repair on top of them to fix any cross-dc inconsistencies.
> 
> Users can always run cross-dc repair to fix the inconsistency after the node operation, but it is too late. That's why we want to make the node operation itself safe without requiring extra repair.
> 
> Using the logic "rebuild from quorum node" is just ignoring the fact we allow users to use R + W > RF and assuming users always use R=quorum and W= quorum. Alternatively, if we fallback to use all or more than one DC, it is slower but it is safer.

I'm not sure I understand what you mean.
Do you suggest that in RBNO we should revert to using all nodes in remove and replace node?

Reply:
I am saying "using from quorum node" is not enough while it brings a lot of complexity in the code.

 For the RF=3, W=1, R=3 , losing one node case, or RF=5, W=2, r=4, losing 2 nodes case. The only safer way for replace is to use nodes in other DCs as well, otherwise, we could break consistency even if user use R+ W > RF strictly. 

Using all nodes is excessive. One trade-off we could make is using all nodes of a dc that has no node down + nodes in local dc.

Reply:
> Using all nodes is excessive. One trade-off we could make is using all nodes of a dc that has no node down + nodes in local dc.

We don't have such logic today.  We can add it as a follow-up optimization.
Right now we have customers that got into trouble when they replaced a node in a DC that lost all its nodes, leaving no replicas behind.  We need to fix that quickly.

Reply:
We can add a config option for replace and a nodetool/api option for remove node to specify the source_dc for RBNO.  This way the admin can make the decision to override the safe defaults and pick another DC to repair from.

Reply:
> > Using all nodes is excessive. One trade-off we could make is using all nodes of a dc that has no node down + nodes in local dc.
> 
> We don't have such logic today. We can add it as a follow-up optimization. Right now we have customers that got into trouble when they replaced a node in a DC that lost all its nodes, leaving no replicas behind. We need to fix that quickly.

Yes, but the fix could be much simpler if we do not add the "using from quorum node" logic which will be removed in the follow-up optimization.

Reply:
So in what conditions would you use the source_dc? when would you return an error? when would you fallback to "another dc + remaining nodes in the source dc?

Reply:
fixed logic in v6

Reply:
> So in what conditions would you use the source_dc? when would you return an error? when would you fallback to "another dc + remaining nodes in the source dc?

Benny, this is what I my mind, to be extra safe:

*) For replace
-  Pick nodes in another dc with all nodes up + remaining nodes in local dc
- If there is no dc with all nodes up, use all nodes in the cluster

*) For rebuild
- if the user provided dc has all nodes up, use it
- if the user provided dc has node down, reject it and recommend another dc with all nodes up. Use the provided dc  anyway if user provides an option to force to rebuild from the dc.

Reply:
@asias, the direction looks good overall, but it disregards the replication factor.
Even if all nodes in the local dc are up other than the replaced or rebuilt node, but rf=1, we have lost the token ranges that were owned by that node, so nodes in another dc should be included as per the above.